File: apps/ios/Sources/Litter/Models/ExperimentalFeatures.swift
import Foundation
import Observation

enum LitterFeature: String, CaseIterable, Identifiable {
    case realtimeVoice = "realtime_voice"
    case ipc = "ipc"
    case generativeUI = "generative_ui"

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .realtimeVoice: return "Realtime"
        case .ipc: return "IPC"
        case .generativeUI: return "Generative UI"
        }
    }

    var description: String {
        switch self {
        case .realtimeVoice: return "Show the realtime voice launcher on the home screen."
        case .ipc: return "Attach to desktop IPC over SSH for faster sync, approvals, and resume. Requires reconnecting the server."
        case .generativeUI: return "Show interactive widgets, diagrams, and charts inline in conversations. Requires starting a new thread."
        }
    }

    var defaultEnabled: Bool {
        switch self {
        case .realtimeVoice: return true
        case .ipc: return true
        case .generativeUI: return false
        }
    }
}

@Observable
final class ExperimentalFeatures {
    static let shared = ExperimentalFeatures()

    @ObservationIgnored private let key = "litter.experimentalFeatures"
    private var overrides: [String: Bool]

    private init() {
        overrides = UserDefaults.standard.dictionary(forKey: key) as? [String: Bool] ?? [:]
    }

    private func persistOverrides() {
        UserDefaults.standard.set(overrides, forKey: key)
    }

    func isEnabled(_ feature: LitterFeature) -> Bool {
        overrides[feature.rawValue] ?? feature.defaultEnabled
    }

    func setEnabled(_ feature: LitterFeature, _ value: Bool) {
        var map = overrides
        if value == feature.defaultEnabled {
            map.removeValue(forKey: feature.rawValue)
        } else {
            map[feature.rawValue] = value
        }
        overrides = map
        persistOverrides()
    }

    func ipcSocketPathOverride() -> String? {
        isEnabled(.ipc) ? nil : ""
    }
}

File: apps/android/app/src/main/java/com/litter/android/ui/ExperimentalFeatures.kt
package com.litter.android.ui

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue

enum class LitterFeature(
    val id: String,
    val displayName: String,
    val description: String,
    val defaultEnabled: Boolean,
) {
    REALTIME_VOICE(
        id = "realtime_voice",
        displayName = "Realtime",
        description = "Show the realtime voice launcher on the home screen.",
        defaultEnabled = true,
    ),
    IPC(
        id = "ipc",
        displayName = "IPC",
        description = "Attach to desktop IPC over SSH for faster sync, approvals, and resume. Requires reconnecting the server.",
        defaultEnabled = true,
    ),
    GENERATIVE_UI(
        id = "generative_ui",
        displayName = "Generative UI",
        description = "Show interactive widgets, diagrams, and charts inline in conversations. Requires starting a new thread.",
        defaultEnabled = false,
    ),
}

object ExperimentalFeatures {
    private const val PREFS = "litter_ui_prefs"
    private const val KEY = "litter.experimentalFeatures"

    private var overrides by mutableStateOf<Map<String, Boolean>>(emptyMap())

    fun initialize(context: Context) {
        @Suppress("UNCHECKED_CAST")
        overrides = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getStringSet(KEY, emptySet())
            ?.mapNotNull { entry ->
                val separator = entry.indexOf('=')
                if (separator <= 0 || separator >= entry.lastIndex) {
                    null
                } else {
                    val featureId = entry.substring(0, separator)
                    val value = entry.substring(separator + 1).toBooleanStrictOrNull() ?: return@mapNotNull null
                    featureId to value
                }
            }
            ?.toMap()
            ?: emptyMap()
    }

    fun isEnabled(feature: LitterFeature): Boolean {
        return overrides[feature.id] ?: feature.defaultEnabled
    }

    fun setEnabled(context: Context, feature: LitterFeature, enabled: Boolean) {
        val next = overrides.toMutableMap()
        if (enabled == feature.defaultEnabled) {
            next.remove(feature.id)
        } else {
            next[feature.id] = enabled
        }
        overrides = next.toMap()
        persist(context)
    }

    fun ipcSocketPathOverride(): String? {
        return if (isEnabled(LitterFeature.IPC)) {
            null
        } else {
            ""
        }
    }

    private fun persist(context: Context) {
        val encoded = overrides.entries.mapTo(linkedSetOf()) { (key, value) -> "$key=$value" }
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putStringSet(KEY, encoded)
            .apply()
    }
}
